### PR TITLE
Add heterogeneous execution path

### PR DESCRIPTION
**Pull request by kurapov-peter**
This is a copy of #333 . I've merged heterogeneous and the default execution paths.

### CODE REVIEW

File: Tests/ArrowBasedExecuteTest.cpp [hunk 1]
```diff
@@ -79,6 +79,8 @@ const size_t g_num_rows{10};
 
 class ExecuteTestBase {
  public:
+  static bool gpusPresent() { return ArrowSQLRunner::gpusPresent(); }
```
Comment (ienkovich) [inline Tests/ArrowBasedExecuteTest.cpp:82]: This is not required and might introduce ambiguity due to `using namespace TestHelpers::ArrowSQLRunner;`

File: Tests/ExecuteTest.cpp [hunk 2]
```diff
@@ -18131,6 +18138,14 @@ int create_and_populate_window_func_table(const bool multi_frag) {
   return 0;
 }
 
+int create_and_populate_window_func_table(const bool multi_frag) {
+  return create_and_populate_table(multi_frag, "test_window_func");
+}
+
+int create_and_populate_multifrag_table() {
```
Comment (ienkovich) [inline Tests/ExecuteTest.cpp:18145]: Unused function?

File: Tests/ExecuteTest.cpp [hunk 3]
```diff
@@ -19381,6 +19396,11 @@ int main(int argc, char** argv) {
                          ->default_value(g_use_groupby_buffer_desc)
                          ->implicit_value(true),
                      "Use GroupBy Buffer Descriptor for hash tables.");
+  desc.add_options()("enable-heterogeneous",
```
Comment (ienkovich) [inline Tests/ExecuteTest.cpp:19399]: I think it would be better to move tests related changes to `ArrowBasedExecuteTest`.

**Review by ienkovich — COMMENTED** [review]
I think the patch is OK overall, but I don't follow tests changes. Also, new files miss the end of the line at the end.

File: Tests/ExecuteTest.cpp (hunk 3 above)
Comment (kurapov-peter) [reply Tests/ExecuteTest.cpp:19399]: Right, these are migration artifacts :)

File: Tests/ArrowBasedExecuteTest.cpp (hunk 1 above)
Comment (kurapov-peter) [reply Tests/ArrowBasedExecuteTest.cpp]: fixed

File: Tests/ExecuteTest.cpp (hunk 2 above)
Comment (kurapov-peter) [reply Tests/ExecuteTest.cpp]: removed

File: Tests/ExecuteTest.cpp [hunk 4]
```diff
@@ -17919,6 +17920,13 @@ TEST(Select, ParseIntegerExceptions) {
   }
 }
 
+TEST(Heterogeneous, Simplest) {
+  if (!g_enable_heterogeneous_execution) {
+    GTEST_SKIP();
+  }
+  c("SELECT count(x) from test_heterogen;", ExecutorDeviceType::GPU);
```
[the remark this answers is not in the document]
Comment (kurapov-peter) [reply Tests/ExecuteTest.cpp]: removed

**Review by alexbaden — APPROVED** [review]

**Pull request merged**